ntoskrnl/io/pnpmgr/pnpnotify.c

```diff
@@ -14,215 +14,309 @@
-        }
-        case EventCategoryHardwareProfileChange:
-        {
-            PHWPROFILE_CHANGE_NOTIFICATION NotificationInfos;
-            NotificationStructure = NotificationInfos = ExAllocatePoolWithTag(
-                PagedPool,
-                sizeof(HWPROFILE_CHANGE_NOTIFICATION),
-                TAG_PNP_NOTIFY);
-            if (!NotificationInfos)
-            {
-                KeReleaseGuardedMutex(&PnpNotifyListLock);
-                return;
-            }
-            NotificationInfos->Version = 1;
-            NotificationInfos->Size = sizeof(HWPROFILE_CHANGE_NOTIFICATION);
-            RtlCopyMemory(&NotificationInfos->Event, Event, sizeof(GUID));
-            break;
-        }
-        case EventCategoryTargetDeviceChange:
-        {
-            if (Event != &GUID_PNP_CUSTOM_NOTIFICATION)
-            {
-                PTARGET_DEVICE_REMOVAL_NOTIFICATION NotificationInfos;
-                NotificationStructure = NotificationInfos = ExAllocatePoolWithTag(
-                    PagedPool,
-                    sizeof(TARGET_DEVICE_REMOVAL_NOTIFICATION),
-                    TAG_PNP_NOTIFY);
-                if (!NotificationInfos)
-                {
-                    KeReleaseGuardedMutex(&PnpNotifyListLock);
-                    return;
-                }
-                NotificationInfos->Version = 1;
-                NotificationInfos->Size = sizeof(TARGET_DEVICE_REMOVAL_NOTIFICATION);
-                RtlCopyMemory(&NotificationInfos->Event, Event, sizeof(GUID));
-            }
-            else
-            {
-                PTARGET_DEVICE_CUSTOM_NOTIFICATION NotificationInfos;
-                NotificationStructure = NotificationInfos = ExAllocatePoolWithTag(
-                    PagedPool,
-                    sizeof(TARGET_DEVICE_CUSTOM_NOTIFICATION),
-                    TAG_PNP_NOTIFY);
-                if (!NotificationInfos)
-                {
-                    KeReleaseGuardedMutex(&PnpNotifyListLock);
-                    return;
-                }
-                RtlCopyMemory(NotificationInfos, EventCategoryData1, sizeof(TARGET_DEVICE_CUSTOM_NOTIFICATION));
-            }
-            break;
-        }
-        default:
-        {
-            DPRINT1("IopNotifyPlugPlayNotification(): unknown EventCategory 0x%x UNIMPLEMENTED\n", EventCategory);
-            KeReleaseGuardedMutex(&PnpNotifyListLock);
-            return;
-        }
+        return;
     }
 
-    /* Loop through procedures registred in PnpNotifyListHead
-     * list to find those that meet some criteria.
-     */
-    ListEntry = PnpNotifyListHead.Flink;
-    while (ListEntry != &PnpNotifyListHead)
+    // the notification may be unregistered inside the procedure
+    // thus reference the driver so we may proceed
+    ObReferenceObject(Entry->DriverObject);
+
+    // release the lock because the notification routine has to be called without any
+    // limitations regarding APCs
+    KeReleaseGuardedMutex(Lock);
```

Comment:
On that note, we should also do what Win7+ does with `IoUnregisterPlugPlayNotificationEx` and guarantee that no callbacks are running (or will run) after the unregister call returns.
This is safe to do for the non-Ex version as well (since it doesn't specify this behavior), and is simply the better architecture.

Reply:
Here we go, now there is `PiReferencePnpNotifyEntry`. Now after the register call returns we may have two situations: either it frees the notification structure (and no callbacks can be run after that obviously) or we're in the middle of a callback and cleanup will occur upon its return. I guess, this matches `IoUnregisterPlugPlayNotificationEx` behavior